engine/src/main/java/io/zeebe/engine/nwe/WorkflowInstanceStateTransitionGuard.java

```diff
@@ -0,0 +1,197 @@
+
+  public WorkflowInstanceStateTransitionGuard(final BpmnStateBehavior stateBehavior) {
+    this.stateBehavior = stateBehavior;
+  }
+
+  /**
+   * Checks if a workflow instance event can be processed based on the current state.
+   *
+   * @return {@code true} if the transition is valid.
+   */
+  public boolean isValidStateTransition(final BpmnElementContext context) {
+    final var result = checkStateTransition(context);
+
+    // log the reason for better debugging
+    result.ifLeft(
+        violation ->
+            LOGGER.debug(
+                "Don't process event because of an illegal state transition: {} [context: {}]",
+                violation,
+                context));
+
+    return result.isRight();
+  }
+
+  private Either<String, ?> checkStateTransition(final BpmnElementContext context) {
+    switch (context.getIntent()) {
+      case ELEMENT_ACTIVATING:
+      case ELEMENT_ACTIVATED:
+      case ELEMENT_COMPLETING:
+      case ELEMENT_COMPLETED:
+        return hasElementInstanceWithState(context, context.getIntent())
+            .flatMap(ok -> hasActiveFlowScopeInstance(context));
+
+      case ELEMENT_TERMINATING:
+      case ELEMENT_TERMINATED:
+        return hasElementInstanceWithState(context, context.getIntent());
+
+      case EVENT_OCCURRED:
+        return hasElementInstanceWithState(context, WorkflowInstanceIntent.ELEMENT_ACTIVATED);
+
+      case SEQUENCE_FLOW_TAKEN:
+        return hasActiveFlowScopeInstance(context);
+
+      default:
+        return Either.left(
+            String.format(
+                "Expected event to have a workflow instance intent but was '%s'",
+                context.getIntent()));
+    }
+  }
+
+  private Either<String, ElementInstance> getElementInstance(final BpmnElementContext context) {
+    final var elementInstance = stateBehavior.getElementInstance(context);
+    if (elementInstance != null) {
+      return Either.right(elementInstance);
+
+    } else {
+      return Either.left(
+          String.format(
+              "Expected element instance with key '%d' to be present in state but not found.",
+              context.getElementInstanceKey()));
+    }
+  }
+
+  private Either<String, ElementInstance> getFlowScopeInstance(final BpmnElementContext context) {
+    final var flowScopeInstance = stateBehavior.getFlowScopeInstance(context);
+    if (flowScopeInstance != null) {
+      return Either.right(flowScopeInstance);
+
+    } else {
+      return Either.left(
+          String.format(
+              "Expected flow scope instance with key '%d' to be present in state but not found.",
+              context.getElementInstanceKey()));
+    }
+  }
+
+  private Either<String, ElementInstance> hasElementInstanceInState(
+      final ElementInstance elementInstance, final WorkflowInstanceIntent expectedState) {
+    final var currentState = elementInstance.getState();
+    if (currentState != expectedState) {
+      return Either.left(
```

Comment:
What is the difference between our old Tuple class and now this either class ? :sweat_smile:

Reply:
`Either` is a way more powerful :rocket: 
It has basically nice functions like `map()` and `flatMap()` that allows chaining the method calls and make it more readable.